AllJoyn/Platform/Bridge/ScriptAdapterLib/JsAdapter.cpp

```diff
@@ -166,139 +184,127 @@ namespace AdapterLib
             RequestPtr = nullptr;
         }
 
         // sanity check
-        AdapterProperty* tempProperty = dynamic_cast<AdapterProperty*>(Property.get());
-        AdapterValue* tempValue = dynamic_cast<AdapterValue*>(Value.get());
-        if (tempValue == nullptr ||
-            tempProperty == nullptr ||
-            tempProperty->Attributes().empty())
+        if (Value == nullptr ||
+            Property == nullptr)
         {
             return static_cast<uint32_t>(ER_BAD_ARG_1);
         }
 
         // find corresponding attribute
-        for (auto attribute : tempProperty->Attributes())
+        for (auto attribute : Property->Attributes())
         {
             if (attribute->Value() != nullptr &&
-                attribute->Value()->Name() == tempValue->Name())
+                attribute->Value()->Name() == Value->Name())
             {
-                attribute->Value()->Data() = tempValue->Data();
+                attribute->Value()->Data() = Value->Data();
                 return ERROR_SUCCESS;
             }
         }
 
-        try
-        {
-            // Run the script to get that property. Script function is "setPROPNAME" where PROPNAME is the name of the property
-            // e.g. getBrightness
-
-            AdapterValueVector paramsIn;
-            AdapterValueVector paramsOut;
-            std::string name("set");
-            name += tempProperty->Name();
-            ScriptHost* host = reinterpret_cast<ScriptHost*>((void*)device->HostContext());
-            std::shared_ptr<IAdapterValue> sharedVal(dynamic_cast<IAdapterValue*>(tempValue));
-            paramsIn.push_back(sharedVal);
-
-            host->CallScriptFunction(name, paramsIn, paramsOut);
-            status = ERROR_SUCCESS;
-        }
-        catch (...)
+        // Run the script to set that property.
+
+        AdapterValueVector paramsIn;
+        AdapterValueVector paramsOut;
+        ScriptHost* host = reinterpret_cast<ScriptHost*>((void*)device->HostContext());
+        paramsIn.push_back(Value);
+
+        auto adapterRequest = std::make_shared<JsAdapterRequest>();
+        RequestPtr = adapterRequest;
+        host->InvokePropertySetter(Property->Name(), Value, [=](uint32_t status)
         {
-            std::cerr << "GetPropertyValue failed on adapter" << std::endl;
-            status = static_cast<uint32_t>(ER_FAIL);
-        }
+            if (status != ERROR_SUCCESS)
+            {
+                std::cerr << "GetPropertyValue failed on adapter" << std::endl;
+            }
 
-        return status;
+            adapterRequest->SetStatus(status);
+        });
+
+        return adapterRequest->Status();
     }
 
     _Use_decl_annotations_
     uint32_t
     JsAdapter::CallMethod(
-		std::shared_ptr<IAdapterDevice> device,
-		std::shared_ptr<IAdapterMethod>& method,
-		std::shared_ptr<IAdapterIoRequest>& /*requestPtr*/
+        std::shared_ptr<IAdapterDevice> device,
+        std::shared_ptr<IAdapterMethod>& method,
```

Comment:
Why? There's no async use of the method reference, so that one is fine. Changing the requestPtr to be passed by value would avoid the need to make an explicit copy on line 286. I'll have to change the IAdapter interface then.